pkg/controller/federatedcluster/controller_test.go

```diff
@@ -0,0 +1,106 @@
+limitations under the License.
+*/
+
+package federatedcluster
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kubernetes-sigs/federation-v2/pkg/apis/core/common"
+	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+)
+
+func TestThresholdCheckedClusterStatus(t *testing.T) {
+
+	epoch := metav1.Now()
+	t1 := metav1.Time{Time: epoch.Add(1 * time.Second)}
+	t2 := metav1.Time{Time: epoch.Add(2 * time.Second)}
+	t3 := metav1.Time{Time: epoch.Add(3 * time.Second)}
+	t4 := metav1.Time{Time: epoch.Add(4 * time.Second)}
+	t5 := metav1.Time{Time: epoch.Add(5 * time.Second)}
+
+	config := util.ClusterHealthCheckConfig{
+		PeriodSeconds:    10 * time.Second,
+		FailureThreshold: 3,
+		SuccessThreshold: 1,
+		TimeoutSeconds:   3 * time.Second,
+	}
+
+	testCases := map[string]struct {
+		clusterStatus         *fedv1a1.FederatedClusterStatus
+		storedClusterData     *ClusterData
+		expectedClusterStatus *fedv1a1.FederatedClusterStatus
+	}{
+		"ClusterReadyAtBegining": {
+			clusterStatus:         clusterStatus(corev1.ConditionTrue, t1, t1),
+			storedClusterData:     &ClusterData{clusterStatus: nil, resultRun: 0},
+			expectedClusterStatus: clusterStatus(corev1.ConditionTrue, t1, t1),
+		},
+		"ClusterNotReadyAtBegining": {
+			clusterStatus:         clusterStatus(corev1.ConditionFalse, t1, t1),
+			storedClusterData:     &ClusterData{clusterStatus: nil, resultRun: 0},
+			expectedClusterStatus: clusterStatus(corev1.ConditionFalse, t1, t1),
+		},
+		"ClusterNotReadyButWithinFailureThreshold": {
+			clusterStatus: clusterStatus(corev1.ConditionFalse, t3, t3),
+			storedClusterData: &ClusterData{
+				clusterStatus: clusterStatus(corev1.ConditionTrue, t2, t1),
+				resultRun:     2},
+			expectedClusterStatus: clusterStatus(corev1.ConditionTrue, t3, t1),
+		},
+		"ClusterNotReadyAndCrossedFailureThreshold": {
+			clusterStatus: clusterStatus(corev1.ConditionFalse, t4, t4),
+			storedClusterData: &ClusterData{
+				clusterStatus: clusterStatus(corev1.ConditionTrue, t3, t1),
+				resultRun:     3},
+			expectedClusterStatus: clusterStatus(corev1.ConditionFalse, t4, t4),
+		},
+		"ClusterReturnToReadyState": {
+			clusterStatus: clusterStatus(corev1.ConditionTrue, t5, t5),
+			storedClusterData: &ClusterData{
+				clusterStatus: clusterStatus(corev1.ConditionFalse, t4, t1),
+				resultRun:     1},
+			expectedClusterStatus: clusterStatus(corev1.ConditionTrue, t5, t5),
+		},
+	}
+
+	for testName, tc := range testCases {
+		t.Run(testName, func(t *testing.T) {
+			newClusterStatus := thresholdAdjustedClusterStatus(tc.clusterStatus, tc.storedClusterData, config)
+			if !reflect.DeepEqual(tc.expectedClusterStatus, newClusterStatus) {
+				t.Fatalf("Unexpected state, expected: %v, got:%v", tc.expectedClusterStatus, newClusterStatus)
+			}
```

Comment:
i guess it's better to check if resultRun is updated as expected.

Reply:
Agree. Added a check in tests.